### PR TITLE
Fix nc21 issues

### LINKED ISSUE #1404: Installation of bookmarks app failed on fresh installed Nextcloud V21
**Describe the bug**
Installation is not possible. even with occ. 
Seems the Tables are not created.
I also tryed to import the blank tables of my old install (v16). But the existing tables are not updated.

**To Reproduce**
Install app bookmarks

**Desktop :**
 - OS: Archlinux
 - Browser: Firefox
 - Version: 85
 
**Server :**
 - OS: Archlinux
 - HTTP server: nginx/1.18.0
 -  PHP 8.0.1 (cli) 
 - Database: mysqlnd 8.0.1 
 - Nextcloud 21.0.0 beta8
 - Bookmarks app version: v4.0.8 Latest

**occ log**
on Installation with occ app:install bookmarks
```
An exception occurred while executing a query: 
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'nextcloud.oc_bookmarks_shared_folders' doesn't exist

```

**Nextcloud log (nextcloud/data/nextcloud.log)**

Installation with webinterface

```
{"reqId":"NzICADt7ZVefmJfjCre0","level":3,"time":"2021-02-05T15:01:30+00:00","remoteAddr":"xx.xx.xx.xx":"handleDriverException","class":"Doctrine\\DBAL\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Connection.php","line":1040,"function":"convertExceptionDuringQuery","class":"Doctrine\\DBAL\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/DB/Connection.php","line":226,"function":"executeQuery","class":"Doctrine\\DBAL\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php","line":210,"function":"executeQuery","class":"OC\\DB\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/DB/QueryBuilder/QueryBuilder.php","line":286,"function":"execute","class":"Doctrine\\DBAL\\Query\\QueryBuilder","type":"->"},
{"file":"/usr/share/webapps/nextcloud/apps/bookmarks/lib/Migration/DeduplicateSharedFoldersRepairStep.php","line":44,"function":"execute","class":"OC\\DB\\QueryBuilder\\QueryBuilder","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Repair.php","line":116,"function":"run","class":"OCA\\Bookmarks\\Migration\\DeduplicateSharedFoldersRepairStep","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/legacy/OC_App.php","line":1054,"function":"run","class":"OC\\Repair","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Installer.php","line":166,"function":"executeRepairSteps","class":"OC_App","type":"::"},
{"file":"/usr/share/webapps/nextcloud/apps/settings/lib/Controller/AppSettingsController.php","line":448,"function":"installApp","class":"OC\\Installer","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":218,"function":"enableApps","class":"OCA\\Settings\\Controller\\AppSettingsController","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":127,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/App.php","line":157,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Route/Router.php","line":302,"function":"main","class":"OC\\AppFramework\\App","type":"::"},
{"file":"/usr/share/webapps/nextcloud/lib/base.php","line":993,"function":"match","class":"OC\\Route\\Router","type":"->"},
{"file":"/usr/share/webapps/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Driver/API/MySQL/ExceptionConverter.php","Line":47,"Previous":{"Exception":"Doctrine\\DBAL\\Driver\\PDO\\Exception","Message":"SQLSTATE[42S02]: Base table or view not found: 1146 Table 'nextcloud.oc_bookmarks_shared_folders' doesn't exist","Code":1146,"Trace":[{"file":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Driver/PDO/Connection.php","line":87,"function":"new","class":"Doctrine\\DBAL\\Driver\\PDO\\Exception","type":"::"},
{"file":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Connection.php","line":1035,"function":"query","class":"Doctrine\\DBAL\\Driver\\PDO\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/DB/Connection.php","line":226,"function":"executeQuery","class":"Doctrine\\DBAL\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php","line":210,"function":"executeQuery","class":"OC\\DB\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/DB/QueryBuilder/QueryBuilder.php","line":286,"function":"execute","class":"Doctrine\\DBAL\\Query\\QueryBuilder","type":"->"},
{"file":"/usr/share/webapps/nextcloud/apps/bookmarks/lib/Migration/DeduplicateSharedFoldersRepairStep.php","line":44,"function":"execute","class":"OC\\DB\\QueryBuilder\\QueryBuilder","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Repair.php","line":116,"function":"run","class":"OCA\\Bookmarks\\Migration\\DeduplicateSharedFoldersRepairStep","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/legacy/OC_App.php","line":1054,"function":"run","class":"OC\\Repair","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Installer.php","line":166,"function":"executeRepairSteps","class":"OC_App","type":"::"},
{"file":"/usr/share/webapps/nextcloud/apps/settings/lib/Controller/AppSettingsController.php","line":448,"function":"installApp","class":"OC\\Installer","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":218,"function":"enableApps","class":"OCA\\Settings\\Controller\\AppSettingsController","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":127,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/App.php","line":157,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Route/Router.php","line":302,"function":"main","class":"OC\\AppFramework\\App","type":"::"},
{"file":"/usr/share/webapps/nextcloud/lib/base.php","line":993,"function":"match","class":"OC\\Route\\Router","type":"->"},
{"file":"/usr/share/webapps/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Driver/PDO/Exception.php","Line":26,"Previous":{"Exception":"PDOException","Message":"SQLSTATE[42S02]: Base table or view not found: 1146 Table 'nextcloud.oc_bookmarks_shared_folders' doesn't exist","Code":"42S02","Trace":[{"file":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Driver/PDO/Connection.php","line":82,"function":"query","class":"PDO","type":"->"},
{"file":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Connection.php","line":1035,"function":"query","class":"Doctrine\\DBAL\\Driver\\PDO\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/DB/Connection.php","line":226,"function":"executeQuery","class":"Doctrine\\DBAL\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php","line":210,"function":"executeQuery","class":"OC\\DB\\Connection","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/DB/QueryBuilder/QueryBuilder.php","line":286,"function":"execute","class":"Doctrine\\DBAL\\Query\\QueryBuilder","type":"->"},
{"file":"/usr/share/webapps/nextcloud/apps/bookmarks/lib/Migration/DeduplicateSharedFoldersRepairStep.php","line":44,"function":"execute","class":"OC\\DB\\QueryBuilder\\QueryBuilder","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Repair.php","line":116,"function":"run","class":"OCA\\Bookmarks\\Migration\\DeduplicateSharedFoldersRepairStep","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/legacy/OC_App.php","line":1054,"function":"run","class":"OC\\Repair","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Installer.php","line":166,"function":"executeRepairSteps","class":"OC_App","type":"::"},
{"file":"/usr/share/webapps/nextcloud/apps/settings/lib/Controller/AppSettingsController.php","line":448,"function":"installApp","class":"OC\\Installer","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":218,"function":"enableApps","class":"OCA\\Settings\\Controller\\AppSettingsController","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":127,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/AppFramework/App.php","line":157,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},
{"file":"/usr/share/webapps/nextcloud/lib/private/Route/Router.php","line":302,"function":"main","class":"OC\\AppFramework\\App","type":"::"},
{"file":"/usr/share/webapps/nextcloud/lib/base.php","line":993,"function":"match","class":"OC\\Route\\Router","type":"->"},
{"file":"/usr/share/webapps/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/usr/share/webapps/nextcloud/3rdparty/doctrine/dbal/src/Driver/PDO/Connection.php","Line":82}},
"CustomMessage":"--"},
"userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:85.0) Gecko/20100101 Firefox/85.0","version":"21.0.0.15"}

```

### PR DESCRIPTION
**Pull request by marcelklehr**
Signed-off-by: Marcel Klehr <mklehr@gmx.net>

fixes #1404

### CODE REVIEW

File: lib/Db/TagMapper.php [hunk 1]
```diff
@@ -96,7 +95,6 @@ public function findByBookmark(int $bookmarkId): array {
 	/**
 	 * @param $userId
 	 * @param string $tag
-	 * @return Statement|int
```
Comment (stijink) [inline lib/Db/TagMapper.php:99]: Why did you remove the return annotations in this file ?

File: lib/Db/Types.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+namespace OCA\Bookmarks\Db;
+
+/**
+ * Database types supported by Nextcloud's DBs
+ *
+ * Use these constants instead of \Doctrine\DBAL\Types\Type or \Doctrine\DBAL\Types\Types
+ *
+ * FIXME: Remove this as soon as min nc version is 21
+ */
+final class Types {
```
Comment (stijink) [inline lib/Db/Types.php:32]: Why do you need to have this types in your own Code and not use it from the core's code ?

File: lib/Db/TagMapper.php (hunk 1 above)
Comment (marcelklehr) [reply lib/Db/TagMapper.php:99]: Nextcloud 21 introduced their own Interface for this, but I want to still support nextcloud 20.

File: lib/Db/Types.php (hunk 2 above)
Comment (marcelklehr) [reply lib/Db/Types.php:32]: Nextcloud 21 updated their Doctrine\DBAL dependency which no longer ships the Types, so Nextcloud 21 ships them as part of their own API, which we can't use until we drop support for Nextcloud 20. So, until then, we'll ship them ourselves.

**Comment by marcelklehr** [comment]
Ignoring Upgrade tests as stable branch is not yet compatible with NC21

File: lib/Db/Types.php (hunk 2 above)
Comment (stijink) [reply lib/Db/Types.php:32]: Ah, i see :-)

**Review by stijink — APPROVED** [review]

**Pull request merged**